### PR TITLE
Fix Battlebond Partner distribution

### PR DESCRIPTION
**Pull request by Sedictious**
This aims to fix the problem @rystan brought up in #5015. Since building on top of the existing generators would affect either readability or efficiency I created an entirely different generator.

I asked a bit around and I think I got the more important rules down:

- There is a maximum of a partner pair per pack

- A pack always contains both partners

- The second partner always takes up an uncommon slot

I am still pretty new (and bad) in Java, so please point out the inevitable stylistic and/or logical mistakes.

### CODE REVIEW

File: Mage/src/main/java/mage/cards/ExpansionSet.java [hunk 1]
```diff
@@ -171,7 +175,36 @@ public String toString() {
...
     }
+    
+    protected int PartnerCheck(List<Card> booster, boolean partnerAllowed, int max, int i){
+
+        for (Ability ability:booster.get(booster.size() - 1).getAbilities()){
+                    //Check if fetched card has the PartnerWithAbility
+                    if (ability instanceof PartnerWithAbility) {               
+                        //Check if the pack already contains a partner pair
+                        if (partnerAllowed){
+                            //Added card always replaces a common card
```
Comment (NoahGleason) [inline Mage/src/main/java/mage/cards/ExpansionSet.java]: I'm guessing this word should be "uncommon"?

Comment (Sedictious) [reply Mage/src/main/java/mage/cards/ExpansionSet.java]: Ahh yes sorry for this, let me fix it

File: Mage.Sets/src/mage/sets/Battlebond.java [hunk 2]
```diff
@@ -9,23 +9,27 @@
...
-
-    private Battlebond() {
+    
+    
+    public Battlebond() {
         super("Battlebond", "BBD", ExpansionSet.buildDate(2018, 6, 8), SetType.SUPPLEMENTAL);
         this.blockName = "Battlebond";
         this.hasBasicLands = false;
         this.hasBoosters = true;
+        this.PartnerMechanic = true;
```
Comment (theelk801) [inline Mage.Sets/src/mage/sets/Battlebond.java]: small nitpick, this should be hasPartnerMechanic or something similar

Comment (Sedictious) [reply Mage.Sets/src/mage/sets/Battlebond.java]: Fixed!

File: Mage.Sets/src/mage/sets/Battlebond.java [hunk 3]
```diff
@@ -9,23 +9,27 @@
...
+    
     private static final Battlebond instance = new Battlebond();
-
+    public int i;
+    
     public static Battlebond getInstance() {
         return instance;
     }
-
-    private Battlebond() {
```
Comment (theelk801) [inline Mage.Sets/src/mage/sets/Battlebond.java]: This constructor should remain private as it is intended to be a singleton class.

File: Mage.Sets/src/mage/sets/Battlebond.java [hunk 4]
```diff
@@ -9,23 +9,27 @@
...
+    
+    
+    public Battlebond() {
```
Comment (theelk801) [inline Mage.Sets/src/mage/sets/Battlebond.java]: The constructor should remain private as it is a singleton class.

File: Mage.Sets/src/mage/sets/Battlebond.java [hunk 5]
```diff
@@ -9,23 +9,27 @@
  * @author TheElk801
  */
 public final class Battlebond extends ExpansionSet {
-
+    
     private static final Battlebond instance = new Battlebond();
-
+    public int i;
```
Comment (theelk801) [inline Mage.Sets/src/mage/sets/Battlebond.java]: What is this variable for?

Comment (Sedictious) [reply Mage.Sets/src/mage/sets/Battlebond.java]: Oops, removed it. Both the variable and the change to public were leftovers from the previous iteration. At first I was going to implement a pattern searching in the set class but after looking closely at the code, decided against it for obvious reasons

File: Mage/src/main/java/mage/cards/ExpansionSet.java [hunk 6]
```diff
@@ -171,7 +175,36 @@ public String toString() {
 
         return theBooster;
     }
+    
+    protected int PartnerCheck(List<Card> booster, boolean partnerAllowed, int max, int i){
```
Comment (ingmargoudt) [inline Mage/src/main/java/mage/cards/ExpansionSet.java]: method should not start with a capital letter
what is int i ?

File: Mage/src/main/java/mage/cards/ExpansionSet.java [hunk 7]
```diff
@@ -171,7 +175,36 @@ public String toString() {
...
+                            }
+                            else{
+                                booster.set(0, card);
+                            }
+                            //2 return value indicates found partner
+                            return 2;
+                        }
 
+                        else{
+                            //If partner already exists, remove card and loop again
```
Comment (ingmargoudt) [inline Mage/src/main/java/mage/cards/ExpansionSet.java]: comment says : loop again, but the scope returns a value.

Comment (Sedictious) [reply Mage/src/main/java/mage/cards/ExpansionSet.java]: You are right in that the comment might be a bit misleading as the looping technically happens outside the function. Essentially the returned value indicates that the maximum number of allowed partners is reached and thus another card must be generated.

File: Mage/src/main/java/mage/cards/ExpansionSet.java (hunk 6 above)
Comment (Sedictious) [reply Mage/src/main/java/mage/cards/ExpansionSet.java]: i is used for checking whether a card can be replaced or safely added (for example consider the case where a partner is the first card, say Khorvath Brightflame, Card1, Card2... without the check the algorithm would replace the first card with its partner Sylvia Brightspear , Card1, Card2). The solution might be too hackish, so I can further work on readability. Would it be better to replace the two parameters with a single boolean?

**Comment by Sedictious** [comment]
@theelk801 I fixed the branch conflict that resulted from one of the recent PRs. Mind taking a look?

**Comment by theelk801** [comment]
Sorry for not responding earlier, I haven't gotten a chance to look over this yet.

**Comment by Sedictious** [comment]
@theelk801 No worries, take your time! Just tell me if there is anything you would like me to improve on.

**Pull request merged**